Add selects

mbucknell opened the pull request:
This does not complete the task. I have brought in Handlebars and am using requirejs.text to load templates as needed. Put the majority of the html that was in index.jsp and put it in a template, home.html. This will template will have pieces removed as we do the sub-views.

Added a BaseView which the majority of our views should extend. Right now it provides a default render which is run during initialize. Possibly we will add common functions (like routing or perhaps common dom manipulation functions) to this as needed.

Used the BaseView to create a HomeView which right now creates the MapView. Eventually other subviews will get created here.

Brought in squire to do the dependancy injection in the tests. Added tests for BaseView and HomeView.

mbucknell commented:
Shout out to @isuftin for providing good documentation of using requirejs and Backbone here: http://isuftin.roughdraft.io/172cf87a4db494269e17-write-testable-backbonejs-applications-using-requirejs

isuftin commented on src/test/js/specs/views/HomeViewSpec.js:
```diff
@@ -0,0 +1,39 @@
+		beforeEach(function(done) {
+			var injector = new Squire();
+			mapViewInitializeSpy = jasmine.createSpy('mapViewInitializeSpy');
+			injector.mock('views/MapView', function() {
```
Ahhh you got the pattern!

isuftin commented on src/main/webapp/js/views/BaseView.js:
```diff
@@ -0,0 +1,65 @@
+		/**
+		 * @constructs
+		 * @param {Object} options
+		 *		@prop router {Backbone.Router instance} - defaults to null
```
Weird tabbin'

mbucknell replied:
Probably because I cut and paste this from github. I'll fix it.

isuftin commented on src/main/webapp/js/templates/home.html:
```diff
@@ -0,0 +1,103 @@
+
+	<!--======Region Search======-->
+	<div id="region-search-container">
+			<p>Find Area of Interest</p>
```
Weird tabbin'

isuftin commented on src/main/webapp/js/templates/home.html:
```diff
@@ -0,0 +1,103 @@
+
+	<!--======Map Filter Sidebar======-->
+	<div id="map-sidebar-container">
+			<div id="filter-container">
```
2 tabs?

mbucknell replied:
I just cut and pasted this from index.jsp, but I can fix.

isuftin commented on src/main/webapp/js/controller/AppRouter.js:
```diff
@@ -1,22 +1,21 @@
+			this.on("route:homeView", function () {
+				this.LOG.trace("Routing to home view");
+
+				this.currentView = new HomeView({
```
Do you think there's any advantage to having the view not render itself on init and, rather, leave that up to the calling function? (this function here)

mbucknell replied:
I have had one or two cases where I didn't want to render the content when the view was created. In those cases, I extended Backbone.View rather than BaseView.  So if we took it out, it would be more flexible. I could go either way.

isuftin replied:
I'd suggest with all else being equal, I personally prefer the pattern of leaving it up to the calling function to do things like render. This also allows for easier testability in some situations.

isuftin commented on src/main/webapp/js/views/BaseView.js:
```diff
@@ -0,0 +1,65 @@
+	var view = Backbone.View.extend({
+
+		LOG : logger.init(),
+		/**
```
Your jsdoc is out of control. (In a good way). Are you autogenning this or are you just that good at it? Do you have  a reference that you use? For example, I would not know how to write proper @returns params and/or what's acceptable. How are you knowing it?

mbucknell replied:
I use this http://usejsdoc.org/ as a reference. As far as what is acceptable, I follow the examples in the doc, but when it comes to something like the @returns below, I use what makes sense to me. No autogeneration, although netbeans sometimes tries to "help". I find the documentation step as well as the testing step frequently make me think about what the code should be doing.

isuftin commented on pom.xml:
```diff
@@ -129,6 +129,7 @@
 		<version.require>2.1.18</version.require>
 		<version.require.text>2.0.14</version.require.text>
 		<version.sinon>1.7.3</version.sinon>
+		<version.squire>0.1.0</version.squire>
```
Looks like James Ward has 0.2.0 up now. That man does not sleep.

mbucknell replied:
He closed the issue at 12:08AM

mbucknell commented:
Incorporated @isuftin comments and removed render from BaseView initialization. Still not sure I understand how use squire effectively but the tests work.

isuftin commented:
As long as they work, they work! :)
